### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by thomasjpfan** [dnouri/skorch#322]
After running `python examples/benchmarks/mnist.py`, I noticed the performance drop compared to pure pytorch.

To investigate this issue, I ran `cProfile` on just the skorch part of the `mnist.py` benchmark. You can reproduce my experiment by following [this gist](https://gist.github.com/thomasjpfan/d078c7086d4e6858be6dd29937f708d7).

Here is the timings for the `fit_loop` function using `skorch`:

<img width="1624" alt="screen shot 2018-08-22 at 9 58 47 pm" src="https://user-images.githubusercontent.com/5402633/44500472-96129a80-a656-11e8-8bad-b80264df32d9.png">

1. The dataloader took 1s in the pure pytorch implementation, and 2.5s in skorch. This is due to the multi-indexing supported by skorch.
2. There is additional overhead to calling `self.optimizer_.step(step_fn)` in `NeutralNet.train_step` with a closure. In the pure pytorch approach the step took 0.6s and in skorch 3s.

**Comment by benjamin-work** [comment]
Thank you for this analysis. We are aware that skorch adds a certain overhead and that will never be zero.

Regarding `multi_indexing`, we have already made huge progress with #233. I'm not sure how much more we can optimize, given that we want to support all those different data types. But we're open to suggestions.

Regarding the closure, the comparison is maybe not apples to apples since the pure torch example doesn't use a closure and hence is less flexible. On the other hand, when we changed skorch to always use closures, we didn't consider performance. So maybe we should make it possible to not use the closure in skorch and avoid the possible overhead. (A quick test did not reveal any performance gain for me, not sure why.)

Overall, the overhead is okay IMO -- the scripts took 11.7 s and 9.9 s, respectively -- but if we can do more without sacrificing too much in code readability and simplicity, we should do that. Also, I would expect the relative overhead to go down when more complex network architectures are used (the example net basically has only 5 layers).
